Add compatibility guide for 1.5 (#2130)
JetBrains/kotlin-web-site: docs/topics/compatibility-guide-15.md

> **Issue**: [KT-27825](https://youtrack.jetbrains.com/issue/KT-27825)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>
> **Short summary**: Kotlin 1.5 will outlaw non-abstract classes containing abstract members invisible from that classes (internal/package-private)
>
> **Deprecation cycle**:
>
> - < 1.5: introduce warning for the problematic classes
> - \>= 1.5: raise this warning to an error,
>  `-XXLanguage:-ProhibitInvisibleAbstractMethodsInSuperclasses` can be used to temporarily revert to pre-1.5 behavior

### Forbid using array based on non-reified type parameters as reified type arguments on JVM

> **Issue**: [KT-31227](https://youtrack.jetbrains.com/issue/KT-31227)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>
> **Short summary**: Kotlin 1.5 will outlaw using array based on non-reified type parameters as reified type arguments on JVM
>
> **Deprecation cycle**:
>
> - < 1.5: introduce warning for the problematic calls
> - \>= 1.5: raise this warning to an error,
>  `-XXLanguage:-ProhibitNonReifiedArraysAsReifiedTypeArguments` can be used to temporarily revert to pre-1.5 behavior

> ### Forbid using array based on non-reified type parameters as reified type arguments on JVM

Reviewer: Duplicates the previous one. Please remove

> **Issue**: [KT-35870](https://youtrack.jetbrains.com/issue/KT-35870)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>
> **Short summary**: Kotlin 1.5 will outlaw secondary enum class constructors which do not delegate to the primary constructor
>
> **Deprecation cycle**:
>
> - < 1.5: introduce warning for the problematic constructors
> - \>= 1.5: raise this warning to an error,
>  `-XXLanguage:-RequiredPrimaryConstructorDelegationCallInEnums` can be used to temporarily revert to pre-1.5 behavior

### Forbid exposing anonymous types from private inline functions

> **Issue**: [KT-33917](https://youtrack.jetbrains.com/issue/KT-33917)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>
> **Short summary**: Kotlin 1.5 will outlaw exposing anonymous types from private inline functions
>
> **Deprecation cycle**:
>
> - < 1.5: introduce warning for the problematic constructors
> - \>= 1.5: raise this warning to an error,
>  `-XXLanguage:-ApproximateAnonymousReturnTypesInPrivateInlineFunctions` can be used to temporarily revert to pre-1.5 behavior

> ### Use correct arguments execution order in calls with named vararg

Reviewer: ```suggestion
### Use the correct order of arguments execution in calls with named vararg
```
Seems easier to read

> **Issue**: [KT-17691](https://youtrack.jetbrains.com/issue/KT-17691)
>
> **Component**: Kotlin/JVM
>
> **Incompatible change type**: behavioral
>

> > **Short summary**: Kotlin 1.5 will change arguments execution order in calls with named vararg

Reviewer: ```suggestion
> **Short summary**: Kotlin 1.5 will change the order of arguments execution in calls with named vararg
```

> **Issue**: [KT-33917](https://youtrack.jetbrains.com/issue/KT-33917)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>
> **Short summary**: Kotlin 1.5 will outlaw exposing anonymous types from private inline functions
>
> **Deprecation cycle**:
>
> - < 1.5: introduce warning for the problematic constructors
> - \>= 1.5: raise this warning to an error,
>  `-XXLanguage:-ApproximateAnonymousReturnTypesInPrivateInlineFunctions` can be used to temporarily revert to pre-1.5 behavior

### Support special semantics for underscore-named catch block parameters

> **Issue**: [KT-31567](https://youtrack.jetbrains.com/issue/KT-31567)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>
> **Short summary**: Kotlin 1.5 will outlaw references to the underscore symbol (`_`) that is used to omit parameter name of an exception in the catch block
>
> **Deprecation cycle**:
>
> - 1.4.20: introduce warning for the problematic references
> - \>= 1.5: raise this warning to an error,
>  `-XXLanguage:-ForbidReferencingToUnderscoreNamedParameterOfCatchBlock` can be used to temporarily revert to pre-1.5 behavior

> ### It's possible to pass non-spread arrays after arguments with SAM-conversion

Reviewer: ```suggestion
### Forbid passing non-spread arrays after arguments with SAM-conversion
```

Reviewer: For consistency with similar issues above

> **Issue**: [KT-35224](https://youtrack.jetbrains.com/issue/KT-35224)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>
> **Short summary**: Kotlin 1.5 will outlaw passing non-spread arrays after arguments with SAM-conversion
>
> **Deprecation cycle**:
>
> - 1.3.70: introduce warning for the problematic calls
> - \>= 1.5: raise this warning to an error,
>  `-XXLanguage:-ProhibitVarargAsArrayAfterSamArgument` can be used to temporarily revert to pre-1.5 behavior

### Change implementation strategy of SAM conversion from anonymous class-based to invokedynamic

> **Issue**: [KT-44912](https://youtrack.jetbrains.com/issue/KT-44912)
>
> **Component**: Kotlin/JVM
>
> **Incompatible change type**: behavioral
>
> **Short summary**: Since Kotlin 1.5, implementation strategy of SAM (single abstract method) conversion will be changed from generating an anonymous class to using the `invokedynamic` JVM instruction
>
> **Deprecation cycle**:
>
> - 1.5: change implementation strategy of SAM conversion,
>  `-Xsam-conversions=class` can be used to revert implementation scheme to the one that used before

> ### Generate nullability assertion for delegated property with generic call in the delegate expression

Reviewer: ```suggestion
### Generate nullability assertion for delegated properties with a generic call in the delegate expression
```

> **Issue**: [KT-44304](https://youtrack.jetbrains.com/issue/KT-44304)
>
> **Component**: Kotlin/JVM
>
> **Incompatible change type**: behavioral
>

> > **Short summary**: Since Kotlin 1.5, the Kotlin compiler will emit nullability assertion for delegated property with generic call in the delegate expression

Reviewer: ```suggestion
> **Short summary**: Since Kotlin 1.5, the Kotlin compiler will emit nullability assertions for delegated properties with a generic call in the delegate expression
```

>
> **Deprecation cycle**:
>
> - 1.5: emit nullability assertion for delegated properties (see details in the issue),
>  `-Xuse-old-backend` or `-language-version 1.4` can be used to temporarily revert to pre-1.5 behavior

> ### Turning warnings into errors for calls with type parameters annotated by @OnlyInputTypes

Reviewer: ```suggestion
### Turn warnings into errors for calls with type parameters annotated by @OnlyInputTypes
```

> **Issue**: [KT-45861](https://youtrack.jetbrains.com/issue/KT-45861)
>
> **Component**: Core language
>
> **Incompatible change type**: source
>

> > **Short summary**: Kotlin 1.5 will outlaw calls like `contains`, `indexOf` and `assertEquals` with senseless arguments to improve type safety (see details in the issue)

Reviewer: ```suggestion
> **Short summary**: Kotlin 1.5 will outlaw calls like `contains`, `indexOf`, and `assertEquals` with senseless arguments to improve type safety
```
I think "see details in the issue" applies to all the changes in this guide :)

### Straighten Char-to-code and Char-to-digit conversions out

> **Issue**: [KT-23451](https://youtrack.jetbrains.com/issue/KT-23451)
>
> **Component**: kotlin-stdlib
>
> **Incompatible change type**: source
>
> **Short summary**: Since Kotlin 1.5, conversions of Char to number types will be deprecated
>
> **Deprecation cycle**:
>
> - 1.5: deprecate `Char.toInt()/toShort()/toLong()/toByte()/toDouble()/toFloat()` and the reverse functions like `Int.toChar()`, and propose replacement 

### Inconsistent case-insensitive comparison of characters in kotlin.text functions

> **Issue**: [KT-45496](https://youtrack.jetbrains.com/issue/KT-45496)
>
> **Component**: kotlin-stdlib
>
> **Incompatible change type**: behavioral
>
> **Short summary**: Since Kotlin 1.5, `Char.equals` will be improved in case-insensitive case by first comparing whether the uppercase variants of characters are equal, then whether the lowercase variants of those uppercase variants (as opposed to the characters themselves) are equal
>
> **Deprecation cycle**:
>
> - < 1.5: old behavior (see details in the issue)
> - 1.5: change behavior for `Char.equals` function 

> ### Remove default locale-dependent case conversion API

Reviewer: ```suggestion
### Remove default locale-sensitive case conversion API
```

> **Issue**: [KT-43023](https://youtrack.jetbrains.com/issue/KT-43023)
>
> **Component**: kotlin-stdlib
>
> **Incompatible change type**: source
>

> > **Short summary**: Since Kotlin 1.5, default locale-dependent case conversion functions like `String.upperCase()` will be deprecated

Reviewer: ```suggestion
> **Short summary**: Since Kotlin 1.5, default locale-sensitive case conversion functions like `String.upperCase()` will be deprecated
```

>
> **Short summary**: Kotlin 1.5 will use default value of the parameter in operator calls
>
> **Deprecation cycle**:
>
> - < 1.5: old behavior (see details in the issue)
> - \>= 1.5: behavior changed,
>  `-XXLanguage:-JvmIrEnabledByDefault` can be used to temporarily revert to pre-1.5 behavior

### Produce empty reversed progressions in for loops if regular progression is also empty 

> **Issue**: [KT-42533](https://youtrack.jetbrains.com/issue/KT-42533)
>
> **Component**: Kotlin/JVM
>
> **Incompatible change type**: behavioral
>
> **Short summary**: Kotlin 1.5 will produce empty reversed progressions in for loops if regular progression is also empty
>
> **Deprecation cycle**:
>
> - < 1.5: old behavior (see details in the issue)
> - \>= 1.5: behavior changed,
>  `-XXLanguage:-JvmIrEnabledByDefault` can be used to temporarily revert to pre-1.5 behavior

### Straighten Char-to-code and Char-to-digit conversions out

> **Issue**: [KT-23451](https://youtrack.jetbrains.com/issue/KT-23451)
>
> **Component**: kotlin-stdlib
>
> **Incompatible change type**: source
>
> **Short summary**: Since Kotlin 1.5, conversions of Char to number types will be deprecated
>
> **Deprecation cycle**:
>

> > - 1.5: deprecate `Char.toInt()/toShort()/toLong()/toByte()/toDouble()/toFloat()` and the reverse functions like `Int.toChar()`, and propose replacement 

Reviewer 2: We've decided to leave `Int.toChar` not deprecated, so let's choose another example, e.g. `Long.toChar` or `Double.toChar`.
